processing/src/main/java/io/druid/segment/DimensionMerger.java

```diff
@@ -0,0 +1,137 @@
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment;
+
+import java.io.IOException;
+import java.nio.IntBuffer;
+import java.util.List;
+
+/**
+ * Processing related interface
+ *
+ * A DimensionMerger is a per-dimension stateful object that encapsulates type-specific operations and data structures
+ * used during the segment merging process (i.e., work done by IndexMerger/IndexMergerV9).
+ *
+ * This object is responsible for:
+ *   - merging any relevant structures from the segments (e.g., encoding dictionaries)
+ *   - writing the merged column data and any merged indexing structures (e.g., dictionaries, bitmaps) to disk
+ *
+ * At a high level, the index merging process can be broken down into the following steps:
+ *   - Merge any value representation metadata across segments
+ *     E.g. for dictionary encoded Strings, each segment has its own unique space of id->value mappings.
+ *     These need to be merged across segments into a shared space of dictionary mappings.
+ *
+ *   - Merge the rows across segments into a common sequence of rows
+ *
+ *   - After constructing the merged sequence of rows, build any applicable index structures (e.g, bitmap indexes)
+ *
+ *   - Write the value representation metadata (e.g. dictionary), the sequence of row values,
+ *     and index structures to a merged segment.
+ *
+ * A class implementing this interface is expected to be highly stateful, updating its internal state as these
+ * functions are called.
+ */
+public interface DimensionMerger
+{
+  /**
+   * Given a list of segment adapters:
+   * - Read any value metadata (e.g., dictionary encoding information) from the adapters
+   * - Merge this value metadata and update the internal state of the implementing class.
+   *
+   * The implementer should maintain knowledge of the "index number" of the adapters in the input list,
+   * i.e., the position of each adapter in the input list.
+   *
+   * This "index number" will be used to refer to specific segments later
+   * in convertSegmentRowValuesToMergedRowValues().
+   *
+   * Otherwise, the details of how this merging occurs and how to store the merged data is left to the implementer.
+   *
+   * @param adapters List of adapters to be merged.
+   * @throws IOException
+   */
+  public void mergeValueMetadataAcrossSegments(List<IndexableAdapter> adapters) throws IOException;
```

Comment:
`writeValueMetadata` is probably a decent name for this. When I first looked at the current name, I was thinking, "where do the merged values go?"